source/server/listener_manager_impl.cc

```diff
@@ -866,7 +875,14 @@ void ListenerManagerImpl::stopListeners(StopListenersType stop_listeners_type) {
         stats_.listener_stopped_.inc();
         for (auto& listener : active_listeners_) {
           if (listener->listenerTag() == listener_tag) {
-            listener->listenSocketFactory().closeAllSockets();
+            if (!listener->udpListenerConfig().has_value() ||
+                listener->udpListenerConfig()->listenerFactory().isTransportConnectionless()) {
+              // Close the listen socket right away to avoid leaving TCP connections in accept queue
+              // already waiting for long timeout. However, connection-oriented UDP listener
+              // shouldn't close the socket because it needs to receive packets on existing
+              // connection via the listen socket.
+              listener->listenSocketFactory().closeAllSockets();
+            }
```

Comment:
nit: Since we have this (almost) identical block of logic twice, could we have a method like `maybeCloseAllSockets(ListenerImplPtr listener)` which wraps this?

Reply:
done